src/PythonClient/src/quixstreams/dataframes/sdf/pipeline.py

```diff
@@ -0,0 +1,186 @@
+        if self._parent:
+            self._graph[self._parent] = self._id
+        else:
+            self._graph[self._id] = None
+
+    def _check_child_contains_parent(self, parent: Self):
+        for key in parent._graph:
+            if parent_child := parent._graph[key]:
+                try:
+                    assert self._graph[key] == parent_child
+                except AssertionError:
+                    raise InvalidFilter
+
+    def _remove_redundant_functions(self, parent: Self):
+        self._check_child_contains_parent(parent)
+        self._functions = self._functions[-(len(self._functions) - len(parent._functions)):]
+        return self
+
+    def _filter(self, item: Column | Self) -> RowApplier:
+        if isinstance(item, Column):
+            return lambda row: row if item.eval(row) else None
+        return lambda row: row if item.process(row) else None
+
+    @staticmethod
+    def _subset(keys: list) -> RowApplier:
+        return lambda row: row._clone(value={k: row.value[k] for k in keys})
+
+    @staticmethod
+    def _set_item(k: str, v: OpValue, row: Row) -> Row:
+        new_row_value = deepcopy(row.value)
+        new_row_value[k] = v
+        return row._clone(value=new_row_value)
+
+    @property
+    def functions(self) -> list[PipelineFunction]:
+        return self._functions
+
+    def _apply(self, func: RowApplier):
+        self._functions.append(PipelineFunction(func=func))
+        return self
+
+    def apply(self, func: Callable[[Row], Optional[dict | Row]]):
+        return self._clone()._apply(func=lambda row: row._clone(value=func(row)))
+
+    def __getitem__(self, item: str | list | Column | Self) -> Column | Self:
+        if isinstance(item, Column):
+            self._apply(self._filter(item))
+            return self
+        elif isinstance(item, list):
+            self._apply(self._subset(item))
+            return self
+        elif isinstance(item, Pipeline):
+            self._apply(self._filter(item._remove_redundant_functions(self)))
+            return self
+        else:
+            return Column(col_name=item)
+
+    def __setitem__(self, key: str, value: Column | OpValue):
+        if isinstance(value, Column):
+            self._apply(lambda row: self._set_item(key, value.eval(row), row))
+        else:
+            self._apply(lambda row: self._set_item(key, value, row))
+
+    @staticmethod
+    def _process_function(func: PipelineFunction, data: Row | List[Row]):
+        if isinstance(data, Row):
+            return func(data)
+        else:
+            data = [func(d) for d in data]
+            return data if any(data) else None
+
+    def process(self, event: Row | List[Row]) -> Optional[Row | List[Row]]:
+        result = event
+        for func in self._functions:
+            print(f'processing func {func._func.__name__}')
+            result = self._process_function(func, result)
+            print(result)
+            if not result:
```

Comment:
I think we should do `is None` check instead of bool. Given that Row has `__bool__` implemented, we might filter out the Row that is has `0` or `""` as a value by mistake.